### PR TITLE
Avoid LINQ in some key methods.

### PR DESCRIPTION
**Pull request by RoosterDragon**
Avoid LINQ in some commonly called Tick methods, and also avoid it when building map domains as this generates a lot of needless work for the GC when loading.

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Player/ConquestVictoryConditions.cs [hunk 1]
```diff
@@ -51,13 +53,17 @@ void ITick.Tick(Actor self)
 			if (!self.Owner.NonCombatant && self.Owner.HasNoRequiredUnits())
 				mo.MarkFailed(self.Owner, objectiveID);
 
-			var others = self.World.Players.Where(p => !p.NonCombatant
-				&& !p.IsAlliedWith(self.Owner));
+			if (otherPlayers == null)
+				otherPlayers = self.World.Players.Where(p => !p.NonCombatant && !p.IsAlliedWith(self.Owner));
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Player/ConquestVictoryConditions.cs]: Neither `NonCombatant` nor `IsAlliedWith` will change after game start, so you could turn this into a smaller concrete collection.

File: OpenRA.Mods.Common/Traits/Player/TechTree.cs [hunk 2]
```diff
@@ -130,13 +130,30 @@ public Watcher(string key, string[] prerequisites, int limit, ITechTreeElement w
 
 			bool HasPrerequisites(Cache<string, List<Actor>> ownedPrerequisites)
 			{
-				return prerequisites.All(p => !(p.Replace("~", "").StartsWith("!") ^ !ownedPrerequisites.ContainsKey(p.Replace("!", "").Replace("~", ""))));
+				// PERF: Avoid LINQ.
+				foreach (var prereq in prerequisites)
+				{
+					var withoutTilde = prereq.Replace("~", "");
+					if (withoutTilde.StartsWith("!") ^ !ownedPrerequisites.ContainsKey(withoutTilde.Replace("!", "")))
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Player/TechTree.cs]: Could you please add `StringComparison.Ordinal` to all these `StartsWith` while you are here?

File: OpenRA.Mods.Common/Traits/Player/ConquestVictoryConditions.cs (hunk 1 above)
Comment (RoosterDragon) [reply OpenRA.Mods.Common/Traits/Player/ConquestVictoryConditions.cs]: Done - also added a comment to that effect for future reference.

File: OpenRA.Mods.Common/Traits/Player/TechTree.cs (hunk 2 above)
Comment (RoosterDragon) [reply OpenRA.Mods.Common/Traits/Player/TechTree.cs]: Added as an extra commit.

**Review by pchote — APPROVED** [review]

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/a1809b89cabd2eef441b5867453fa9f145632b55...208fe2fea6795895f357a42c7f578a0363fdf557

**Pull request merged**